Statistical models ws (#22)
EDUCE-UBC/educer: inst/tutorials/statistical_models_ws/statistical_models_ws.Rmd

---
title: "Statistical models in R"
author: >
  Gil J. B. Henriques and Andrew Li<br>
  Based on notes by Yue Liu and Kim Dill-McFarland.<br>
  A tutorial by [ECOSCOPE](http://ecoscope.ubc.ca/) at UBC.
date: "version `r format(Sys.time(), '%B %d, %Y')`"
description:  >
  Learn to perform statistical analyses with complex data sets.
  You will learn about linear regression, ANOVA, ANCOVA, 

>   mixed effects models, and generalized linear models.

Reviewer: ```suggestion
author: >-
  Gil J. B. Henriques and Andrew Li<br>
  Edited by Stephan Koenig<br>
  Based on notes by Yue Liu and Kim Dill-McFarland<br>
  A tutorial by [ECOSCOPE](http://ecoscope.ubc.ca/) at UBC.
date: "version `r format(Sys.time(), '%B %d, %Y')`"
description:  >-
  Learn to perform statistical analyses with complex data sets.
  You will learn about linear regression, ANOVA, ANCOVA, 
  mixed-effects models, and generalized linear models.
```

> output: 
>   learnr::tutorial:
>       progressive: true
>       allow_skip: true
> runtime: shiny_prerendered

Reviewer: ```suggestion
output: 
  learnr::tutorial:
    progressive: true
    allow_skip: true
    css: /css/style.css
runtime: shiny_prerendered
```

The [tidyverse](https://www.tidyverse.org/) is a collection of R packages. Packages expand R's capabilities by providing additional functions. In this tutorial we will use many functions from tidyverse packages such as dplyr (for data manipulation; pronounced *dee-plier*) and ggplot2 (for data visualization). You can learn more about the tidyverse in our "Introduction to the R tidyverse" tutorial.

When you work in a script that makes heavy use of functions from a particular package, we recommend loading it at the top of your script. You can use the code below to load all of the tidyverse packages:

```{r echo = TRUE, message = FALSE}
library(tidyverse) # Easily install and load the tidyverse packages
```

In this tutorial, we will also occasionally use functions from the following packages: gapminder, datasets, lme4, car, lsmeans, HSAUR. Because we will only make light use of each of these packages, there is no need to load all of the packages up-front at the top of our script. Instead of loading the entire package, we will access only the specific functions we need using the double colon operator, `::`. You will learn more about this operator later in the tutorial.

> Keep in mind that to use packages in a script, these packages *must already be installed* on your computer (see the "Introduction to R and RStudio" tutorial for more details about installing packages and working with scripts). By installing the educer package, all of the packages we will use in this tutorial were automatically installed as well.

Reviewer: ````suggestion
## Introduction

In this tutorial, we will learn about several common statistical models. They all employ a family of techniques called *regression analysis*, where we estimate the relationship between an *independent variable* (also called a *predictor*) and a *dependent variable* (also called a *response* or *outcome*). We will introduce linear regression, ANOVA, ANCOVA, and mixed-effects models for continuous response data, logistic regression binary response data, as well as Poisson and Negative Binomial regression for count response data.

You will learn:

-   The different functions used to build a statistical model in R,
-   The assumptions behind the different models,
-   How the formula object in R is used to specify all the model terms,
-   How to interpret the main effects and interaction terms in a model.

This is an intermediate tutorial that assumes some prior experience with base R (see our "Introduction to R and RStudio" tutorial) and with the tidyverse package collection (see our "Introduction to the R tidyverse" tutorial).

```{r child = system.file("resources/markdown", "how_to_follow_tutorials.Rmd", package = "educer")}
```

### Data description

Unlike our other tutorials, "Statistical Models" utilizes several datasets to demonstrate the use of statistical tests accurately. These datasets are built into base R or R packages. In this tutorial, you will find more information on each of these datasets in its relevant section(s).
 
### Loading packages

The [tidyverse](https://www.tidyverse.org/) is a collection of R packages. Packages expand R's capabilities by providing additional functions. This tutorial will use many functions from tidyverse packages such as dplyr (for data manipulation; pronounced *dee-plier*) and ggplot2 (for data visualization). You can learn more about the tidyverse in our "Introduction to the R tidyverse" tutorial.

When you work in a script that makes heavy use of functions from a particular package, we recommend loading it at the top of your script. You can use the code below to load all of the tidyverse packages:

```{r echo = TRUE, message = FALSE}
library(tidyverse) # Easily install and load the tidyverse packages
```

This tutorial will occasionally use functions from the following packages: gapminder, datasets, lme4, car, lsmeans, HSAUR. Because we will only make light use of each of these packages, there is no need to load all packages up-front at the top of our script. Instead of loading the entire package, we will access only the specific functions we need using the double colon operator, `::`. You will learn more about this operator later in the tutorial.

Remember that to use packages in a script, these packages *must already be installed* on your computer (see the "Introduction to R and RStudio" tutorial for more details about installing packages and working with scripts). By installing the educer package, all of the packages we will use in this tutorial were automatically installed.
````

## Analysis of variance (ANOVA)

ANOVA is used when you have data with:

-   a quantitative response/dependent variable ($Y$) such as:
  -   height
  -   salary
  -   number of offspring
-   one or more categorical explanatory/independent variable(s) ($X$) such as:
  -   eye color
  -   sex
  -   genotype at a given locus

For example, you would use ANOVA to address questions like:

-   Does diet has an effect on weight gain?
  -   response variable: weight gain (e.g., kg)
  -   explanatory variable: type of diet (e.g., low vs. medium vs. high sugar)
-   Does the type sexual relationship practiced influence the fitness of male Red-winged Blackbirds?
  -   response variable: fitness of male bird (e.g., number of eggs laid)

>   -   explanatory variable: sexual relationship (e.g., monagamy vs. polygamy)

Reviewer: ```suggestion
## Analysis of variance (ANOVA)

ANOVA is used when you have data with:

-   a quantitative response/dependent variable ($Y$) such as:
    -   height
    -   salary
    -   number of offspring
-   one or more categorical explanatory/independent variable(s) ($X$) such as:
    -   eye colour
    -   sex
    -   genotype at a given locus

For example, you would use ANOVA to address questions like:

-   Does diet affect weight gain?
    -   response variable: weight gain (e.g., kg)
    -   explanatory variable: type of diet (e.g., low vs. medium vs. high sugar)
-   Does the type of sexual relationship practiced influence the fitness of male Red-winged Blackbirds?
    -   response variable: fitness of male bird (e.g., number of eggs laid)
    -   explanatory variable: sexual relationship (e.g., monogamy vs. polygamy)
```

Always double-check that the formatting is correct after rendering the output/tutorial.

> 

Reviewer: ```

With ANOVA, you will determine to accept or reject the null hypothesis:
-   Null hypothesis $H_0$: population means of all groups are equal
-   Alternate hypothesis $H_1$: population means of at least two groups are not equal

```

Now let's run some ANOVAs on real data! We will start simple, by comparing the mean of two-groups. We will focus first on the case of a single explanatory variable (which is called the *one-way* ANOVA).

There are two perfectly acceptable statistical tests in R that we could apply to compare data between two groups. The first, which you may be very familiar with, is the *t*-test. The second is the ANOVA. Interestingly, the *t*-test is really a special case of ANOVA that can be used when only comparing only two groups. ANOVA is a more general test that we will later see can be used with more than two groups as well as more than one explanatory variable. 

#### Load and explore the data

The first experiment we are going to analyze was done to address the question of whether sexual activity effects the longevity of male fruit flies. To assess this, we are going to use a modified version of the fruitfly data from the [faraway R package](https://cran.r-project.org/web/packages/faraway/faraway.pdf).

Our hypotheses for this experiment are as follows:

Null hypothesis, $H_0$: Sexual activity has no effect on the population mean longevity of male fruit flies.

> Alternative hypothesis, $H_A$: Sexual activity has an effect on population mean longevity of male fruit flies.

Reviewer: ```suggestion
### Key assumptions

-   Observations in the sample are independent of each other.
-   ANOVA is robust to the non-normality of sample data
-   Homoscedasticity — the variance of data in groups should be the same. However, if the ANOVA is balanced (groups have equal sample size), the model is robust to unequal variance.

### The gist of the math

When we run an ANOVA, we calculate a test statistic (*F*-statistic) that compares the **between** group variation with the **within** group variation:

$$F=MSB/MSW$$

where $MSB =$ Mean Square Between (i.e., the mean squared distance from each observation to the sample mean) and $MSW =$ Mean Square Within (i.e., the mean squared distance from each observation to the group mean). Essentially, if there is greater variation between the groups than within the groups, we will get a large test statistic value (and correspondingly, a small *p*-value) and reject that null hypothesis ($H_0$: population means of all groups are equal).

If you want to delve into ANOVA in more depth, check out [this video tutorial](https://www.khanacademy.org/math/statistics-probability/analysis-of-variance-anova-library).

### One-way ANOVA with two groups

Now let's run some ANOVAs on real data! We will start simple by comparing the mean of the two groups. We will focus first on the case of a single explanatory variable (which is called the *one-way* ANOVA).

There are two perfectly acceptable statistical tests in R that we could apply to compare data between two groups. The first, which you may be very familiar with, is the *t*-test. The second is the ANOVA. Interestingly, the *t*-test is a special case of ANOVA that can be used when only comparing two groups. ANOVA is a more general test that we will later see can be used with more than two groups as well as more than one explanatory variable. 

#### Load and explore the data

The first experiment we are going to analyze was done to address whether sexual activity affects the longevity of male fruit flies. To assess this, we will use a modified version of the fruitfly data from the [faraway R package](https://cran.r-project.org/web/packages/faraway/faraway.pdf).

Our hypotheses for this experiment are as follows:
-   Null hypothesis $H_0$: Sexual activity does not affect the population mean longevity of male fruit flies.
-   Alternative hypothesis $H_A$: Sexual activity affects the population mean longevity of male fruit flies.
```

> Let's now load the `fruitfly` dataset from the faraway package and assign it to a variable called `df_fly`. We haven't loaded this package, so we have to explicitly write the name of the package in front of the dataset's name (separated by a double colon operator: `faraway::fruitfly`), so that R knows where to find it. Run the code box below to load and inspect the first few rows of the data. You will have to replace the `<placeholder>` by the name of the variable you have created:

Reviewer: ```suggestion
Let's now load the `fruitfly` dataset from the faraway package and assign it to a variable called `df_fly`. We haven't loaded this package, so we have to explicitly write the name of the package in front of the dataset's name (separated by a double colon operator: `faraway::fruitfly`) so that R knows where to find it. Run the code box below to load and inspect the first few rows of the data. You will have to replace the `<placeholder>` by the name of the variable you have created:
```

```{r load-faraway, exercise = TRUE}
df_fly <- faraway::fruitfly
head(<variable_name>)
```

```{r load-faraway-solution}
df_fly <- faraway::fruitfly
head(df_fly)
```

```{r setup-fly, include = FALSE}
df_fly <- faraway::fruitfly
```

The first two columns are continuous variables representing the fruit fly's thorax length and longevity. The final column, `activity`, is a categorical variable that encodes the fly's sexual activity. If the male was kept solitary (`isolated`) or with pregnant females (either `one` or `many`), he was not sexually active. On the other hand, the male may have been kept with virgin females, in which case he was sexually active (`low` = one virgin female; `high` = eight virgin females).

> The next step is to create two-level categorical groups from the numerical variables. We will also subset the data set to equal group sizes (by randomly selecting twenty observations from each group) If you are unfamiliar with the functions below, see our "Introduction to the R tidyverse" tutorial. Use the code box below to create the modified data frame and examine its first few rows:

Reviewer: ```suggestion
The first two columns are continuous variables representing the fruit fly's thorax length and longevity. The final column, `activity`, is a categorical variable that encodes the fly's sexual activity. If the male was kept solitary (`isolated`) or with pregnant females (either `one` or `many`), he was not sexually active. On the other hand, the male may have been kept with virgin females, in which case he was sexually active (`low` = one virgin female; `high` = eight virgin females).

The next step is to create two-level categorical groups from the numerical variables. We will also subset the data set to equal group sizes (by randomly selecting twenty observations from each group). If you are unfamiliar with the functions below, see our "Introduction to the R tidyverse" tutorial. Use the code box below to create the modified data frame and examine its first few rows:
```

 
```{r fly-2-groups, exercise = TRUE, exercise.setup = "setup-fly"}
fly_2_groups <- df_fly %>% 
	# Change "activity" to a 2-level variable
	mutate(activity = ifelse(activity %in% c("isolated", "one", "many"),
														"no", "yes")) %>%
	# Change size to a 2-level variable (for later)
	mutate(thorax = ifelse(thorax <= 0.8, "short", "long")) %>% 
	# Subset to equal group sizes
	group_by(activity, thorax) %>% 
	sample_n(20)

# Examine first few rows of the new data frame
head(<variable_name>)
``` 

```{r fly-2-groups-solution}
fly_2_groups <- df_fly %>% 
	# Change "activity" to a 2-level variable
	mutate(activity = ifelse(activity %in% c("isolated", "one", "many"),
														"no", "yes")) %>%
	# Change size to a 2-level variable (for later)
	mutate(thorax = ifelse(thorax <= 0.8, "short", "long")) %>% 
	# Subset to equal group sizes
	group_by(activity, thorax) %>% 
	sample_n(20)

# Examine first few rows of the new data frame
head(fly_2_groups)
``` 
 
```{r setup-groups, include = FALSE}

> fly_2_groups <- df_fly %>% 

Reviewer: ```suggestion
fly_2_groups <- faraway::fruitfly %>% 
```

You must check each exercise. Otherwise, you won't catch errors. Here the setup chunk did not actually work, and as a result, the next exercises all won't work. Each exercise chunk has its own environment, so variables (like `df_fly` here) are not shared from previous chunks.

	# Change "activity" to a 2-level variable
	mutate(activity = ifelse(activity %in% c("isolated", "one", "many"),
														"no", "yes")) %>%
	# Change size to a 2-level variable (for later)
	mutate(thorax = ifelse(thorax <= 0.8, "short", "long")) %>% 
	# Subset to equal group sizes
	group_by(activity, thorax) %>% 
	sample_n(20)
```
 
To further explore our new data set, use the `dim()` function (to get the number of rows and columns) and the `summary()` function (to get a summary of the data). 

```{r fly-2-explore, exercise = TRUE, exercise.setup = "setup-groups"}
# get number of rows and columns
dim(<variable_name>)

# see summary of data
summary(<variable_name>)
``` 

> Now, we should visualize the data to understand its basic properties, find patterns, and explore modeling strategies. Exploratory graphs are usually made really quickly. We simply want to get a better understanding of the data at this point.

Reviewer: ```suggestion
Now, we should visualize the data to understand its basic properties, find patterns, and explore modelling strategies. Exploratory graphs are usually made really quickly. We simply want to get a better understanding of the data at this point.
```

Modify the code box below to explore the effect of the explanatory variable, `longevity` on the response variable, `activity`:

> ```{r fly-2-eda, exercise = TRUE, excercise.setup = "setup-groups"}

Reviewer: ````suggestion
```{r fly-2-eda, exercise = TRUE, exercise.setup = "setup-groups"}
````
These are basic mistakes that you should catch by running the exercises.

# create a box plot for exploratory data analysis
<variable_name> %>%
  ggplot(aes(x = <explanatory_column>, y = <response_column>)) +
  geom_boxplot() +
  labs(x = "Sexually Active", y = "Longevity (days)")
```

```{r fly-2-eda-solution}
fly_2_groups %>%

>   ggplot(aes(x = longevity, y = activity)) +

Reviewer: ```suggestion
  ggplot(aes(x = activity, y = longevity)) +
```
You also need to check if the provided answers are correct and work. Here the data was plotted wrong.

  geom_boxplot() +
  labs(x = "Sexually Active", y = "Longevity (days)")
```

> 

Reviewer: ```suggestion
```

#### Formula notation in R

To perform an ANOVA in R, we need to understand R's formula notation, as this is the first argument we give to the ANOVA function (`aov()`). The formula notation stars with providing the response variable, then a special character, `~` which can be read as "as a function of", and then the explanatory/independent variable(s). Thus, the formula notation for this experiment is:
```
longevity ~ activity
```
The formula notation can get more complex, such as including additional explanatory/independent variables or interaction terms. We will introduce these as we attempt more complex analyses later on.

#### ANOVA in R
We want to know if sexual activity has an effect on population mean longevity of male fruit flies. 
The `aov()` function can be used to create a model that answers this question. 
As stated above, the first argument to this function is the formula. 
The second argument is the name of the data frame variable, in our case  `fly_2_groups`.

> Create an ANOVA model object called `fly_2_groups_model`:

Reviewer: ````suggestion
To perform an ANOVA in R, we need to understand R's formula notation, as this is the first argument we give to the ANOVA function (`aov()`). The formula notation starts with providing the response variable, then a special character, `~`, which can be read as "as a function of," and then the explanatory/independent variable(s). Thus, the formula notation for this experiment is:

```
longevity ~ activity
```

The formula notation can get more complex, including additional explanatory/independent variables or interaction terms. We will introduce these as we attempt more complex analyses later on.

#### ANOVA in R

We want to know if the sexual activity affects the population mean longevity of male fruit flies. 
The `aov()` function creates a model that answers this question. 
As stated above, the first argument for this function is the formula. 
The second argument is the name of the data frame variable, in our case  `fly_2_groups`.

Create an ANOVA model object called `fly_2_groups_model`. The `aov()` function returns a "model" object. You can examine the results of the model with the `summary()` function:
````

> ```{r two-group, exercise = TRUE}
> # The response variable is on the left and explanatory variable on the right
> fly_2_groups_model <- aov(<response_column> ~ <explanatory_column>,
>                           data = <variable_name>)
> ```
> 
> ```{r two-group-solution}
> fly_2_groups_model <- aov(longevity ~ activity, data = fly_2_groups)
> ```
> 
> The `aov()` function returns a "model" object. You can examine the results of the model with the `summary()` function:
> 
> ```{r two-group-sum, exercise = TRUE}
> summary(<model_object>)
> ```
> 
> ```{r two-group-sum-solution, excercise.setup = "setup-two-group-sum"}
> summary(<fly_2_groups_model>)
> ```
> 

Reviewer: ````suggestion
```{r two-group, exercise = TRUE, exercise.setup = "setup-groups"}
# The response variable is on the left and explanatory variable on the right
fly_2_groups_model <- aov(<response_column> ~ <explanatory_column>,
                          data = <variable_name>)
summary(<model_object>)
```
```{r two-group-solution}
fly_2_groups_model <- aov(longevity ~ activity, data = fly_2_groups)
summary(fly_2_groups_model)
```
````
Only 2 out of 6 exercises have worked so far.

So, what does this output mean? The most important result in regards to rejecting or failing to reject our null hypothesis is the right-most number, the p-value (`Pr(>F)`). In this simple one-way ANOVA, we have a single p-value which has a very small value. Given that this is much smaller then the commonly used threshold for rejecting the null hypothesis, p < 0.05, we can reject our null hypothesis that sexual activity has no effect on the population mean longevity of male fruit flies, and accept the alternative hypothesis that sexual activity **does** have an effect on population mean longevity of male fruit flies. 

> #### **Exercise 1: 1-way ANOVA**
> Using ANOVA, test if fruit fly longevity is affected by size (as measured by thorax length).
> 
>   * What are your null and alternate hypotheses?
>   * What can you conclude from these results?
> 
> Create a model variable called `exercise1` and use the `summary()` function to examine the output.
> 
> **Note: Hints will be provided for exercises but no solution. Try to figure it out!**
> ```{r ex1, exercise = TRUE}
> <model_object> <- aov(<response_column> ~ <explanatory_column>,
>                       data = <variable_name>)
> <function_for_examining_output>(<model_object>)
> ```
> 
> ```{r ex1-hint-1}
> # The response column is thorax and the exploratory column is thorax
> ```
> 
> ```{r ex1-hint-2}
> # Name the model object exercise1!
> ```
> 
> ```{r ex1-hint-3}
> # To examine the output of your model, use the summary() function!
> ```

Reviewer: ````suggestion
::: exercise
**1-way ANOVA**: Using ANOVA, test if fruit fly longevity is affected by size (as measured by thorax length).

  * What are your null and alternate hypotheses?
  * What can you conclude from these results?

Create a model variable called `exercise1` and use the `summary()` function to examine the output.

**Note: Hints will be provided for exercises but no solution. Try to figure it out!**
```{r ex1, exercise = TRUE}
<model_object> <- aov(<response_column> ~ <explanatory_column>,
                      data = <variable_name>)
<function_for_examining_output>(<model_object>)
```

```{r ex1-hint-1}
# The response variable is longevity, and the explanatory variable is thorax.
```

```{r ex1-hint-2}
# Name the model object exercise1!
```

```{r ex1-hint-3}
# To examine the output of your model, use the summary() function!
```
:::

````
I just pushed the necessary `style.css` to use the exercise `<div>` as demonstrated here.

The code chunk above does not work. Please fix.
